lib/winappdriver.js

```diff
@@ -1,36 +1,250 @@
+
+const REQD_PARAMS = ['app'];
+const DEFAULT_HOST = '127.0.0.1';
+const DEFAULT_PORT = "4823"; //  should be non-4723 to avoid conflict on the same box
+
+class WinAppDriver extends events.EventEmitter {
   constructor(opts = {}) {
+    const {proxyhost, proxyport, executable, cmdArgs} = opts;
+    super();
+
     for (let req of REQD_PARAMS) {
       if (!opts || !opts[req]) {
         throw new Error(`Option '${req}' is required!`);
       }
       this[req] = opts[req];
     }
 
-    this.jwproxy = new JWProxy({ server: this.host, port: this.port });
-    this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
+    this.proxyHost = proxyhost || DEFAULT_HOST;
+    this.proxyPort = proxyport || DEFAULT_PORT;
+    this.cmdArgs = cmdArgs;
+    this.winappdriver = executable;
+    this.proc = null;
+    this.executableVerified = false;
+    this.state = WinAppDriver.STATE_STOPPED;
+    this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort});
+}
+
+  async startWinAppDriverProcess (caps) {
+    this.initWinAppDriverPath();
+    return this.start(caps, true);
+  }
+
+  async initWinAppDriverPath () {
+    if (this.executableVerified) return;
+    
+    let binPath = this.winappdriver || (await getWinAppDriverBinaryPath());
+    if (!await fs.exists(binPath)) {
+      throw new Error(`Trying to use a WinAppDriver binary at the path ` +
+                      `${binPath}, but it doesn't exist!`);
+    }
+    this.winappdriverpath = binPath;
+    this.executableVerified = true;
+    log.info(`Set WinAppDriver binary as: ${this.winappdriverpath}`);
+  }
+
+  async start (caps, emitStartingState = true) {
+
+    this.capabilities = caps;
+    if (emitStartingState) {
+      this.changeState(WinAppDriver.STATE_STARTING);
+    }
+
+    // XXXYD TODO: would be better if WinAppDriver didn't require passing in /wd/hub as a param
+    let args = [this.proxyPort + "/wd/hub"];
+
+    if (this.cmdArgs) {
+      args = args.concat(this.cmdArgs);
+    }
+
+    const startDetector = (stdout) => {
+        // XXXYD What is this for, looking for a log entry that shows we're up?
+        // return stdout.indexOf('ENTER') != -1;
+        return true; 
+      };
+
+    let processIsAlive = false;
+        try {
+          await this.initWinAppDriverPath();
+          await this.killAll();
+
+          // set up our subprocess object
+          this.proc = new SubProcess(this.winappdriverpath, args);
+          processIsAlive = true;
+
+          // handle log output
+          this.proc.on('output', (stdout, stderr) => {
+            if (stdout) {
+              log.info(`[STDOUT] ${stdout.trim()}`);
+            }
+            if (stderr) {
+              log.info(`[STDERR] ${stderr.trim()}`);
+            }
+          });
+
+     // handle out-of-bound exit by simply emitting a stopped state
+      this.proc.on('exit', (code, signal) => {
+        processIsAlive = false;
+        if (this.state !== WinAppDriver.STATE_STOPPED &&
```

Comment:
who knows if we want to use the same state machine as for chromedriver. maybe we can get away without 'restarting'?